### PR TITLE
feat(server): add lusca

### PR DESCRIPTION
**Pull request by Awk34**

### CODE REVIEW

**Comment by kingcody** [comment]
Are there any notes that should be included for the users? Or is the more a drop-in/set-it-forget-it kind of feature?

**Comment by Awk34** [comment]
As is, it's kind of a drop-in, but there are more features that can be enabled, etc. I'll add a comment in the code about Lusca.

**Comment by kingcody** [comment]
This needs to be rebased on `canary`.

File: app/templates/server/config/express.js [hunk 1]
```diff
@@ -13,11 +13,12 @@ var methodOverride = require('method-override');
...
 var errorHandler = require('errorhandler');
 var path = require('path');
+var lusca = require('lusca');
 var config = require('./environment');<% if (filters.auth) { %>
-var passport = require('passport');<% } %><% if (filters.twitterAuth) { %>
+var passport = require('passport');<% } %>
 var session = require('express-session');<% if (filters.mongoose) { %>
 var mongoStore = require('connect-mongo')(session);
-var mongoose = require('mongoose-bird')();<% } %><% } %>
+var mongoose = require('mongoose-bird')();<% } %>
```
Comment (kingcody) [inline app/templates/server/config/express.js]: requiring `mongoose` directly

**Comment by Awk34** [comment]
rebased

File: app/templates/server/config/express.js [hunk 2]
```diff
@@ -13,8 +13,9 @@ var methodOverride = require('method-override');
 var cookieParser = require('cookie-parser');
 var errorHandler = require('errorhandler');
 var path = require('path');
+var lusca = require('lusca');
 var config = require('./environment');<% if (filters.auth) { %>
-var passport = require('passport');<% } %><% if (filters.twitterAuth) { %>
+var passport = require('passport');<% } %>
 var session = require('express-session');<% if (filters.mongoose) { %>
 var mongoStore = require('connect-mongo')(session);
 var mongoose = require('mongoose');<% } %><% } %>
```
Comment (kingcody) [inline app/templates/server/config/express.js]: If you're removing `<% if (filters.twitterAuth) { %>` then you also need to remove the closing tag `<% } %>` here.

Comment (Awk34) [inline app/templates/server/config/express.js]: oh damnit, yeah, that's from the rebase

**Comment by kingcody** [comment]
I think the tests are failing because they're making a `POST` with out sending the `CSRF` token.

**Comment by kingcody** [comment]
I rebased this on canary, but I'm going to guess you already figured that out.

**Comment by Awk34** [comment]
> I think the tests are failing because they're making a POST with out sending the CSRF token.

Oh yeah, nice catch!

**Comment by kingcody** [comment]
I'm going to hold off on merging this for now; I want to try to get supertest working with the `CSRF` tokens. I'll work on it tonight, unless someone beats me to it.

**Comment by Awk34** [comment]
> I'm going to hold off on merging this for now; I want to try to get supertest working with the CSRF tokens. I'll work on it tonight, unless someone beats me to it.

What's the point?

**Comment by kingcody** [comment]
If we disable the csrf strategy for the tests then we're not really testing the same code that would run in production. Also it would be good to know if a future change breaks the implementation.

**Comment by Awk34** [comment]
I don't see there really being a difference if you were to get supertest to work with the CSRF tokens. It wouldn't affect the tests still.

**Comment by kingcody** [comment]
They are integration(e2e) tests, which are supposed to test an app as a whole. IMO that includes security implementation.

I'm not trying to test if lusca is using and decoding tokens properly, I'm going to assume it is. What I'd like to test is that our implementation is correct.

**Comment by Awk34** [comment]
If we can get it working with tests, that would be better; but if not, I don't thing it's a *huge* deal.

**Comment by kingcody** [comment]
I agree, I wouldn't want tests to put too much of a hold on implementing a security feature. I'll debug it more this weekend and see if I can get things straight between Travis and SauceLabs. Oddly enough the tests pass when I use `sauce_connect` locally but for some reason Travis is a different situation.

Either way I'm setting Monday as my deadline.

**Comment by kingcody** [comment]
Well there is still something fishy about the way the e2e test fail when using `lusca + travis + sauce_connect` but I was not able to get to the bottom of it. I do wonder if we have some race conditions buried down deep, however I wasn't able to expose them in my tests.

**Pull request merged**